Dynamic Product Creation

ljgilm opened the pull request:
Related to issue #13

ljgilm committed:
Admin Integrated #13

ljgilm committed:
Revert "Admin Integrated #13"

This reverts commit 87f6e00537a2f26c0ef250a6399f858177ad749b.

ljgilm committed:
Added default image #13

ljgilm committed:
Admin Integrated to index #13

ljgilm committed:
Admin Store.js - changed window.onload #13

ljgilm committed:
Admin added http request for client side #13

ljgilm committed:
Admin created scriptMuuri #13

ljgilm committed:
Added admin logic to view and changed window.onload #13

ljgilm committed:
Admin created CSS file #13

ljgilm committed:
Admin app.js #13

ljgilm committed:
Admin server response #13

ljgilm committed:
Admin added admin routes #13

ljgilm committed:
Admin added dependencies to package.json #13

dpaceoffice commented on server/controllers/default-controller.js:
```diff
@@ -117,4 +117,117 @@ export default class Controller {
+            {
+                id: 15,
+                title: "Great Sword",
+                description: "A great sword, used by the largest men",
+                price: 25.0,
+                image: "./assets/greatsword.png"
+            }
+
+        ]
+        //console.log("ARRAY");
+        //console.log(products);
+        response.json(products);
+
+    }
+
+    addProduct(request, response) {
+        response.send(request.body);
+        console.log("POST SUCCESS! Added product! Can update mongo!");
+        console.log(request.body);
+
+        let products = {
+            0: {
+                id: 0,
+                title: "Katana",
+                description: "A very very sharp sword used by samurai",
+                price: 5.0,
+                image: "./assets/katana.png"
+            },
+            1: {
+                id: 1,
+                title: "Great Sword",
+                description: "A great sword, used by the largest men",
+                price: 25.0,
+                image: "./assets/greatsword.png"
+            },
+        }
+        products[Number(request.body.id)] = request.body;
+        //console.log("APPENDED");
+        //console.log(products);
+    }
```
Needs refractoring for mongo

dpaceoffice commented on server/controllers/default-controller.js:
```diff
@@ -117,4 +117,117 @@ export default class Controller {
+
+    }
+
+    productById(request, response) {
+        const id = request.params.id;
+        console.log("ID: " + request.params.id);
+        response.json({
+            id: id,
+            title: "Great Sword - EDIT",
+            description: "A great sword, used by the largest men. GETTING FOR EDIT",
+            price: 25.0,
+            image: "./assets/greatsword.png"
+        })
+    }
```
Needs refactoring for mongo

dpaceoffice commented on server/app.js:
```diff
@@ -21,10 +21,22 @@ else {
 function setupApp() {
     app.use(cors());
     app.use(bodyParser.json());
+    app.use(bodyParser.urlencoded({ extended: false }))     //use body parser to read url 
```
This may cause unintended consequences. We need to test if it'll work on the main branch.

dpaceoffice commented on server/app.js:
```diff
@@ -21,10 +21,22 @@ else {
 const port = process.env.PORT | 3000;
+
+/* ADMINISTRATION CODE */
+function localhostHandler(request, response, next) {
+    response.header('Access-Control-Allow-Origin', '*');
+    response.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
+    response.header('Access-Control-Allow-Methods', 'PUT, POST, GET, DELETE, OPTIONS');
+    next();
+}
```
Possibly unnecessary since we have cors

dpaceoffice commented on client/scripts/view.js:
```diff
@@ -59,3 +59,221 @@ const displayCheckout = (body, total, checkout) => {
+        editData.description = document.getElementById("editProduct-description").value;
+        editData.price = document.getElementById("editProduct-price").value;
+        editData.image = document.getElementById("editProduct-image").value;
+        //console.log(editData);
+        postEditProduct(editData);
+    }
+
+    document.getElementById("saveOrder").addEventListener("click", function () {
+        let data = window.localStorage.getItem('prodIdlayout');
+        postLayout(data);
+    });
+
+    [...document.querySelectorAll('.removeProdBut')].forEach(function (item) {
+        item.addEventListener('click', function () {
+            console.log("REMOVE PRODUCT!");
+        });
+    });
+
+    [...document.querySelectorAll('.editProduct')].forEach(function (item) {
+        item.addEventListener('click', async function () {
+            let prodID = item.getAttribute('product_number')
+            console.log(prodID + " - EDIT PRODUCT! from edit button");
+            let product = await getProductById(prodID)
+            console.log("Product info from server for editting " + product.id);
+            
+            // NEED TO GET PRODUCT'S INFORMATION
+            document.getElementById("editProduct-title").value = product.title;
+            document.getElementById("editProduct-description").value = product.description;
+            document.getElementById("editProduct-price").value = product.price;
+        });
+    });
+
+    document.querySelector('.grid').addEventListener('click', function (e) {
+        if (elementMatches(e.target, '.card-remove, .card-remove i')) {
+            postRemove(e);
+            console.log("sent remove post");
+        }
+    });
+
+}
```
All logic gets handled in controller.js for MVC model

dpaceoffice commented on client/scripts/view.js:
```diff
@@ -59,3 +59,221 @@ const displayCheckout = (body, total, checkout) => {
+        `
+    )
+
+    var exampleModal = document.getElementById('exampleModal');
+    exampleModal.addEventListener('show.bs.modal', function (event) {
+        // clear values
+        document.getElementById("product-title").value = "";
+        document.getElementById("product-description").value = "";
+        document.getElementById("product-price").value = "";
+        document.getElementById("product-image").value = "";
+
+        // Button that triggered the modal
+        var button = event.relatedTarget
+        // Extract info from data-bs-* attributes
+        var recipient = button.getAttribute('data-bs-whatever')
+        // If necessary, you could initiate an AJAX request here and then do the updating in a callback.
+        // Update the modal's content.
+        var modalTitle = exampleModal.querySelector('.modal-title')
+        //modalTitle.textContent = 'New message to ' + recipient
+        modalTitle.textContent = "Add Product";
+
+        const submitButton = document.getElementById('submitProduct');
+        submitButton.addEventListener('click', submitEvent);
+    })
```
All logic gets handled in controller.js for MVC model

dpaceoffice commented on client/scripts/view.js:
```diff
@@ -59,3 +59,221 @@ const displayCheckout = (body, total, checkout) => {
+
+}
+
+window.onload = (event) => {
+    loadAdminPage();
+    //startPage();
+    //document.getElementById("start").addEventListener("click", function () { loadAdminPage()});
+};
```
Admin should have its own model, like Store, and the html page would reference the admin model for onload.

dpaceoffice commented on client/scripts/controller.js:
```diff
@@ -10,4 +10,128 @@ const addController = function (...buttonIDs) {
+    return data;
+}
+
+/*^^^^ ADMINISTRATION CODE ^^^^^*/
```
This should go into a model for admin, the controller passes functionality to the model.

dpaceoffice commented on client/scripts/Store.js:
```diff
@@ -1,5 +1,5 @@
 
-window.onload = init();
+// ADMINISTRATION window.onload = init();
 post = function (url, data) {
     return fetch(url, { method: "POST", headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(data) });
 }
```
unnecessary, see other comments.

dpaceoffice reviewed (changes requested):
Changes must be made in order to follow the MVP model, and some refactoring for mongoose.

ljgilm committed:
Refactored Admin - index.html #13

ljgilm committed:
Refactored Admin - Admin.js Model #13

ljgilm committed:
Refactored Admin - Created Callbacks Controller.js #13

ljgilm committed:
Refactored Admin - refactored to admin.js and controller.js #13

ljgilm committed:
Admin - adjusted muuri for refactoring of admin #13

ljgilm committed:
made onload  call init() again

ljgilm committed:
Refactored Admin - refactored view to controller.js and admin.js #13

ljgilm committed:
Admin - adjusted CSS for refactoring of admin #13

ljgilm committed:
Admin - removed extra code that was from lab #13

ljgilm committed:
Admin - adjustment to add product to mongodb with David #13

ljgilm committed:
Admin - updated package.json #13

ljgilm committed:
Admin Category - see products by Category index.html #13

ljgilm committed:
Revert "Admin Category - see products by Category index.html #13"

This reverts commit 99b204ac759c28a40efb2379c7e8331260895209.

ljgilm committed:
Admin Category - get products by category index.html #13

ljgilm committed:
Admin Category - get products by category Admin.js #13

ljgilm committed:
Admin Category - get products by category controller.js #13

ljgilm committed:
Admin Category - get products by category scriptMuuri.js #13

ljgilm committed:
Admin Category - get products by category view.js #13

ljgilm committed:
Admin Category - get products by category default-controller.js #13

ljgilm committed:
Admin Category - get products by category default-routes.js #13

ljgilm committed:
Admin Category - add, edit, remove, layout, and create cat (admin.js) #13

can add, remove, edit products by category
can save layout by category
can create new category

ljgilm committed:
Admin Category - add, edit, remove, layout, and create cat (controller.js) #13

can add, remove, edit products by category
can save layout by category
can create new category

ljgilm committed:
Removed controller_admin no longer needed - now inside controller.js #13

Admin Category - add, edit, remove, layout, and create cat has been accomplished
can add, remove, edit products by category
can save layout by category
can create new category

ljgilm committed:
Admin Category - add, edit, remove, layout, and create cat (scriptMuuri.js) #13

can add, remove, edit products by category
can save layout by category
can create new category

ljgilm committed:
Admin Category - add, edit, remove, layout, and create cat (view.js) #13

can add, remove, edit products by category
can save layout by category
can create new category

ljgilm committed:
Admin Category - add, edit, remove, layout, and create cat (def-controller.js) #13

can add, remove, edit products by category
can save layout by category
can create new category

ljgilm committed:
Admin Category - delete cat. and display current cat. title Admin.js #13

ljgilm committed:
Admin Category - delete cat. and display current cat. title Controller.js #13

ljgilm committed:
Admin Category - delete cat. and display current cat. title View.js #13

ljgilm committed:
Admin Category - delete cat. and display current cat. title Style.css #13

ljgilm committed:
Admin Category - delete cat. and display current cat. title Default-controller.js #13

ljgilm committed:
Admin Category - delete cat. and display current cat. title Default_routes.js #13

ljgilm committed:
Admin trying to merge: moving client folder into server folder

ljgilm committed:
Admin: changed default-controller to config name
